pkg/cmd/experimental/diagnostics/master.go

```diff
@@ -0,0 +1,143 @@
+	systemddiagnostics "github.com/openshift/origin/pkg/diagnostics/systemd"
+	diagnostictypes "github.com/openshift/origin/pkg/diagnostics/types/diagnostic"
+)
+
+const (
+	MasterDiagnosticsRecommendedName = "master"
+
+	StandardMasterConfigPath string = "/etc/openshift/master/master-config.yaml"
+)
+
+var (
+	AvailableMasterDiagnostics = util.NewStringSet("AnalyzeLogs", "UnitStatus", "MasterConfigCheck")
+)
+
+// user options for openshift-diagnostics client command
+type MasterDiagnosticsOptions struct {
+	RequestedDiagnostics util.StringList
+
+	MasterConfigLocation string
+
+	LogOptions *log.LoggerOptions
+	Logger     *log.Logger
+}
+
+const longMasterDescription = `
+OpenShift Diagnostics
+
+This command helps you understand and troubleshoot a running OpenShift
+master. It is intended to be run from the same context as the master
+(where "openshift start" or "openshift start master" is run, possibly from
+systemd or inside a container) and with the same configuration options.
+
+    $ %s
+`
+
+func NewMasterCommand(name string, fullName string, out io.Writer) *cobra.Command {
+	o := &MasterDiagnosticsOptions{
+		RequestedDiagnostics: AvailableMasterDiagnostics.List(),
+		LogOptions:           &log.LoggerOptions{Out: out},
+	}
+
+	cmd := &cobra.Command{
+		Use:   name,
+		Short: "Troubleshoot an OpenShift v3 master.",
+		Long:  fmt.Sprintf(longMasterDescription, fullName),
+		Run: func(c *cobra.Command, args []string) {
+			kcmdutil.CheckErr(o.Complete())
+
+			failed, err := o.RunDiagnostics()
+			o.Logger.Summary()
+			o.Logger.Finish()
+
+			kcmdutil.CheckErr(err)
+			if failed {
+				os.Exit(255)
+			}
+		},
+	}
+
+	cmd.SetOutput(out) // for output re: usage / help
+
+	cmd.Flags().StringVar(&o.MasterConfigLocation, "master-config", "", "path to master config file")
+	diagnosticflags.BindLoggerOptionFlags(cmd.Flags(), o.LogOptions, diagnosticflags.RecommendedLoggerOptionFlags())
+	diagnosticflags.BindDiagnosticFlag(cmd.Flags(), &o.RequestedDiagnostics, diagnosticflags.NewRecommendedDiagnosticFlag())
+
+	return cmd
+}
+
+func (o *MasterDiagnosticsOptions) Complete() error {
+	// set the master config location if it hasn't been set and we find it in an expected location
+	if len(o.MasterConfigLocation) == 0 {
+		if _, err := os.Stat(StandardMasterConfigPath); !os.IsNotExist(err) {
+			o.MasterConfigLocation = StandardMasterConfigPath
+		}
+
+	}
+
+	var err error
+	o.Logger, err = o.LogOptions.NewLogger()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o MasterDiagnosticsOptions) RunDiagnostics() (bool, error) {
+	diagnostics := map[string]diagnostictypes.Diagnostic{}
+
+	// if we don't have a master config file, then there's no work to do
+	if len(o.MasterConfigLocation) == 0 {
+		// TODO remove MasterConfigCheck from the list
+	}
+
+	systemdUnits := systemddiagnostics.GetSystemdUnits(o.Logger)
+
+	for _, diagnosticName := range o.RequestedDiagnostics {
+		switch diagnosticName {
+		case "AnalyzeLogs":
```

Comment:
Reducing the inputs down the minimum and creating a portable `Diagnostic` interface allows people to easily create the `Diagnostics` themselves.  That's where the flexibility needs to be and where it really helps.

By having individual `Diagnostics` that are built on minimal set of information is possible to navigate both up and down the chain to determine what information is required at each step and why and where they are used.  Linked together it like this, it's clear that the `MasterConfigCheck` needs a file location.  The `MasterConfigCheck` type itself has a minimal set of information and its very obvious what's used an why.  The command Options have a `MasterConfigLocation` because the `MasterConfigCheck` requires it and that drives the need for the flag.  The chain is easily inspectable from one end to the other and there is traceability.